Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+  where
+    -- This fails unless we have --global and --user, unlike
+    -- pkgDoc elsewhere in ghc-mod.
+    toDocDirOpts pkg dbs = ["field", pkg, "haddock-html", "--global", "--user"] ++ ghcPkgDbStackOpts dbs
+
+ghcPkgHaddockInterface
+    :: forall (m :: * -> *).  MonadIO m
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> String
+    -> m (Maybe String)
+ghcPkgHaddockInterface ghcPkg readProc pkgDbStack p = do
+    hout <- liftIO $ readProc ghcPkg (toHaskellInterfaces p pkgDbStack) ""
+    return $ Safe.lastMay $ words $ reverse . dropWhile (== '\n') . reverse $ hout
+  where
+    toHaskellInterfaces pkg dbs = ["field", pkg, "haddock-interfaces", "--global", "--user"] ++ ghcPkgDbStackOpts dbs
+
+getVisibleExports
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => (String -> m (Maybe String))
+    -> String
+    -> m (Maybe (M.Map String [String]))
+getVisibleExports getHaddockInterfaces p = do
+    gmLog GmDebug "" $ strDoc $ "getVisibleExports: " ++ p
+
+    let p' = case splitOn "@" p of
+                [p0, _] -> p0
+                _       -> p
+    
+    haddockInterfaceFile <- getHaddockInterfaces p'
+    join <$> traverse getVisibleExports' haddockInterfaceFile
+
+  where
+
+    getVisibleExports' :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m) => FilePath -> m (Maybe (M.Map String [String]))
+    getVisibleExports' ifile = do
+        iface <- Haddock.readInterfaceFile nameCacheFromGhc ifile
+
+        case iface of
+            Left _          -> do gmErrStrLn $ "Failed to read the Haddock interface file: " ++ ifile
+                                            ++ "You probably installed packages without using the '--enable-documentation' flag."
+                                            ++ ""
+                                            ++ "Try something like:\n\n\tcabal install --enable-documentation p"
+                                  error "No haddock interfaces file, giving up."
+            Right iface'    -> do let m  = map (\ii -> (Haddock.instMod ii, Haddock.instVisibleExports ii)) $ Haddock.ifInstalledIfaces iface' :: [(Module, [Name])]
+                                      m' = map (\(mname, names) -> (showSDoc tdflags $ ppr mname, map (showSDoc tdflags . ppr) names)) m       :: [(String, [String])]
+                                  return $ Just $ M.fromList m'
+
+    ------------------------------------------------------------------------------------------------------------------------
+    -- Copied from http://hackage.haskell.org/package/haddock-api-2.16.1/docs/src/Haddock-InterfaceFile.html#nameCacheFromGhc
+    -- but for a general monad m instead of the specific monad Ghc.
+    ------------------------------------------------------------------------------------------------------------------------
+    nameCacheFromGhc :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m) => Haddock.NameCacheAccessor m
```

Comment:
`GmOut` and `GmLog` seem to be redundant constraints here, you should compile with GHC 8 that warns about those now. Also please don't use `forall` notation if you don't have to.